### PR TITLE
HttpClientMockBuilder supports "send" method

### PR DESCRIPTION
**Pull request by michal-tichy-84**

### CODE REVIEW

File: src/MockeryTools/Http/HttpClientMockBuilder.php [hunk 1]
```diff
@@ -144,4 +219,63 @@ private function createRequestUrl(string $endpoint): string
...
+            if ($requestHeaders[$headerName][0] !== $headerValue) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+
+    private function createRequestMatcher(
```
Comment (dominikvoda) [inline src/MockeryTools/Http/HttpClientMockBuilder.php:240]: I would like to have \Mockery\Matcher\MatcherAbstract here... it's much more elegant ;)

File: src/MockeryTools/Http/HttpClientMockBuilder.php [hunk 2]
```diff
@@ -144,4 +219,63 @@ private function createRequestUrl(string $endpoint): string
     {
         return $this->basePath . $endpoint;
     }
+
+
+    private function containsExpectedHeaders(RequestInterface $request): bool
+    {
+        $requestHeaders = $request->getHeaders();
```
Comment (dominikvoda) [inline src/MockeryTools/Http/HttpClientMockBuilder.php:226]: why it is not possible to compare headers like `$request->getHeaders() === $this->expectedHeaders` ?

File: src/MockeryTools/Http/HttpClientMockBuilder.php [hunk 3]
```diff
@@ -124,16 +130,85 @@ public function expectFailedRequest(
...
+        assert(is_a($guzzleExceptionClassname, GuzzleException::class, true));
+        $expectedRequestBody = Json::encode($expectedRequestData);
+        $request = new Request(
+            $expectedHttpMethod,
+            $this->createRequestUrl($expectedEndpoint),
+            $this->expectedHeaders,
+            $expectedRequestBody
+        );
+        $response = new Response($errorCode, [], Json::encode($responseData));
+        $exceptionToThrow = $this->getExceptionToThrow($guzzleExceptionClassname, $request, $response, $errorCode);
```
Comment (dominikvoda) [inline src/MockeryTools/Http/HttpClientMockBuilder.php:185]: I think we should have same logic for expectRequest and expectSend... we we need to distinguish them?

File: src/MockeryTools/Http/HttpClientMockBuilder.php (hunk 2 above)
Comment (michal-tichy-84) [reply src/MockeryTools/Http/HttpClientMockBuilder.php:226]: There are often some additional headers, that do not concern us

File: src/MockeryTools/Http/HttpClientMockBuilder.php [hunk 4]
```diff
@@ -61,24 +61,24 @@ public function build(): ClientInterface
 
 
     /**
-     * @param mixed[] $expectedResponseBody
-     * @param mixed[] $requestBody
+     * @param array<string, mixed> $responseData
+     * @param array<string, mixed> $expectedRequestData
```
Comment (dominikvoda) [inline src/MockeryTools/Http/HttpClientMockBuilder.php:65]: to bylo spravne, nikdy nevis jak request/response vypada

File: src/MockeryTools/Http/HttpClientMockBuilder.php [hunk 5]
```diff
@@ -88,52 +88,109 @@ public function expectRequest(
...
-            ->andThrow($requestException);
+            ->andThrow($exceptionToThrow);
+
+        return $this;
+    }
+
+
+    /**
+     * @param array<string, mixed> $responseData
+     * @param array<string, mixed> $expectedRequestData
```
Comment (dominikvoda) [inline src/MockeryTools/Http/HttpClientMockBuilder.php:127]: same here

File: src/MockeryTools/Http/HttpClientMockBuilder.php [hunk 6]
```diff
@@ -88,52 +88,109 @@ public function expectRequest(
...
+            ->once()
+            ->andReturn(new Response(200, [], $responseBody));
+
+        return $this;
+    }
+
+
+    /**
+     * @param array<string, mixed> $responseData
+     * @param array<string, mixed> $expectedRequestData
```
Comment (dominikvoda) [inline src/MockeryTools/Http/HttpClientMockBuilder.php:152]: a tady taky

File: src/MockeryTools/Http/HttpClientMockBuilder.php [hunk 7]
```diff
@@ -88,52 +88,109 @@ public function expectRequest(
...
+            ->once()
+            ->andThrows($exceptionToThrow);
 
         return $this;
     }
 
 
     /**
-     * @param mixed[] $requestBody
+     * @param array<string, mixed> $requestData
```
Comment (dominikvoda) [inline src/MockeryTools/Http/HttpClientMockBuilder.php:184]: a tady taky :)

File: src/MockeryTools/Http/HttpClientMockBuilderTest.php [hunk 8]
```diff
@@ -17,6 +23,9 @@ final class HttpClientMockBuilderTest extends TestCase
     private const HEADERS = ['Authorization' => 'Bearer thisIsSomeToken'];
 
 
+    /**
+     * @throws Throwable
+     */
```
Comment (dominikvoda) [inline src/MockeryTools/Http/HttpClientMockBuilderTest.php:28]: uplne zbytecny kod...k cemu to tady je? :)

File: src/MockeryTools/Http/HttpRequestMatcher.php [hunk 9]
```diff
@@ -74,4 +75,20 @@ public function __toString(): string
     {
         return '<HttpRequest:' . $this->expectedUri . '>';
     }
+
+
+    private function containsExpectedHeaders(RequestInterface $request): bool
```
Comment (dominikvoda) [inline src/MockeryTools/Http/HttpRequestMatcher.php:80]: proc nestaci ===?

File: src/MockeryTools/Http/HttpRequestMatcher.php [hunk 10]
```diff
@@ -43,8 +43,12 @@ public function __construct(
 
         $this->expectedMethod = $expectedMethod;
         $this->expectedUri = $expectedUri;
-        $this->expectedHeaders = $expectedHeaders;
         $this->expectedBody = $expectedBody;
+
+        $this->expectedHeaders = [];
+        foreach ($expectedHeaders as $headerName => $headerValue) {
```
Comment (dominikvoda) [inline src/MockeryTools/Http/HttpRequestMatcher.php:49]: mozna bych tady dal komentar `proc?`

**Review by dominikvoda — APPROVED** [review]

File: src/MockeryTools/Http/HttpRequestMatcher.php (hunk 9 above)
Comment (michal-tichy-84) [reply src/MockeryTools/Http/HttpRequestMatcher.php:80]: Protože to nefunguje. Automaticky se tam přidávají headery, které mě nezajímají, třeba Host=api.com

File: src/MockeryTools/Http/HttpClientMockBuilderTest.php (hunk 8 above)
Comment (michal-tichy-84) [reply src/MockeryTools/Http/HttpClientMockBuilderTest.php:28]: Mr. svítící Storma :stuck_out_tongue_winking_eye:

**Pull request merged**